src/Contacts.js

```diff
@@ -0,0 +1,172 @@
+        this.currentContact = this.localData.contacts[this.props.index] ??
+            { name: "",
+            email: "",
+            comment: "" };
+        this.editable = this.props.index == this.props.searchParams.get(contactKey)
+
+
+        makeObservable(this, {
+            currentContact:observable,
+            countryCode:observable,
+            emailError:observable,
+            nameError:observable,
+            editable:observable,
+            localData:observable,
+
+            makeEditable:action.bound,
+            updateName:action.bound,
+            updateEmail:action.bound,
+            updateComment:action.bound,
+            saveChanges:action.bound,
+            cancelChanges:action.bound,
+            deleteContact:action.bound,
+        });
+    }
+
+    makeEditable() {
+        this.props.updateSearchParams({ editingContact: this.props.index })
+        this.editable = true;
+    }
+
+    updateName(event) {
+        this.currentContact.name = event.target.value;
+        if (this.currentContact.name === "") {
+            this.nameError = true;
+        } else { this.nameError = false; }
+    }
+
+    updateEmail(event) {
+        this.currentContact.email = event.target.value;
+
+        if (!this.currentContact.email.includes('@')) {
+            this.emailError = true;
+        } else { this.emailError = false; }
+    }
+
+    updateComment(event) {
+        this.currentContact.comment = event.target.value;
+    }
+
+    saveChanges() {
+        this.editable = false;
+        addOrUpdateContact(this.countryCode, this.props.index, this.currentContact)
+        this.props.updateSearchParams({ editingContact: null })
+    }
+
+    cancelChanges() {
+        this.editable = false;
+        this.props.updateSearchParams({ editingContact: null })
+        if (this.localData.contacts[this.props.index].name === "") {
+            deleteContact(this.countryCode, this.props.index)
+            this.localData = getLocalData(this.countryCode)
+            return
+        }
+
+        this.currentContact.name = this.localData.contacts[this.props.index].name;
+        this.currentContact.email = this.localData.contacts[this.props.index].email;
+        this.currentContact.comment = this.localData.contacts[this.props.index].comment;
+    }
+
+    deleteContact() {
+        deleteContact(this.countryCode, this.props.index);
+        this.localData = getLocalData(this.countryCode)
+    }
+
+    render() {
+        let {
+            emailError,
+            nameError,
+            localData,
+            editable,
+            currentContact,
+            makeEditable,
+            updateName,
+            updateEmail,
+            updateComment,
+            saveChanges,
+            cancelChanges,
+            deleteContact
+        } = this;
+
+        return(
+            <div>
+            {
+                Object.keys(localData.contacts).includes(this.props.index) &&

```

Comment:
also, you shouldnt need this? seems weird that this component would even exist if its index didnt exist in the `localData.contacts` obj

Reply:
I added that in after it was throwing errors from contact deletion. It would be trying to access the object with the deleted index, which would cause an error.